### ISSUE TITLE
changing xcat from to postgres threw away node switch information.

### ISSUE THREAD

**Issue by ralphbellofatto** [xcat2/xcat-core#2037]
While closing out this issue we found a problem with xcat:
https://gitlabhost.rtp.raleigh.ibm.com/4A8491897/research-92-cluster/issues/141

We migrated our xcat db from SQLite to postgres by followed the directions here:

http://xcat-docs.readthedocs.io/en/stable/advanced/hierarchy/databases/postgres_configure.html 

specifically we executed:

```
1) yum install postgresql*
yum install perl-DBD-Pg

2) pgsqlsetup -i -V  (-V will spit out verbose output, looking at trimming this down some..) 

That should automatically migrate the existing SQLite DB to PostgreSQL 

1:59:21 PM: Will look like this: 

[root@fs4 log]# lsxcatd -a
Version 2.12.4 (git commit 2bf96ee97df3e3a04ef72c42f2e756af22ecad5a, built Tue Oct 25 11:15:48 EDT 2016)
This is a Management Node
cfgloc=Pg:dbname=xcatdb;host=192.168.3.25|xcatadm
dbengine=Pg
dbname=xcatdb
dbhost=192.168.3.25
dbadmin=xcatadm

[root@fs4 log]# cat /etc/xcat/cfgloc 
Pg:dbname=xcatdb;host=192.168.3.25|xcatadm|XXXXX
```

When we were done, we compared the output of the lsdef -l command with what it was prior to doing the update and we found the following two fields are missing on every node:

```
switch=c460tors01
switchport=3
```

The following attachment contains the before(orig) and after(new) node definitions extracted by xcat.
[xcat-postgres-migrate.zip](https://github.com/xcat2/xcat-core/files/554187/xcat-postgres-migrate.zip)

**Comment by ralphbellofatto** [comment]
NOTE:
We attempted to restore the above configuration with
cat node.lsdef | chdef -z

And it failed to update the switch ports...

We extracted the data by hand from the lsdef file and then constructed a script as follows:
```
chdef c460c001 switch=c460tors01 switchport=3
chdef c460c002 switch=c460tors01 switchport=4
chdef c460c003 switch=c460tors01 switchport=5
chdef c460c004 switch=c460tors01 switchport=6
chdef c460c005 switch=c460tors01 switchport=7
```

And then the data stuck...

**Comment by ralphbellofatto** [comment]
reopening...

**Comment by whowutwut** [comment]
To add to this, if we create a group with switch/switchport defined in the group, it's not able to be stored into the postgres DB.    The chdef command does not report and error. 
```
[root@fs4 fs4]# chdef -t group -o disco switchport='|\D+(\d+)\D+(\d+)|($2*1)|'
1 object definitions have been created or modified.
[root@fs4 fs4]# lsdef -t group  -o disco -i switchport
Object name: disco
    switchport=
```

@tingtli  Can you review the FVT scenarios for PostGresDB and the test for attributes against the DB?

**Comment by whowutwut** [comment]
This one must be resolved in 2.12.4, adding high priority.

**Comment by immarvin** [comment]
hi @chenglch , would you please take a look at this?

**Comment by chenglch** [comment]
Look at table structure at first, the column `node`, `switch` and `port` are defined as composite keys. On sqlite and mysql db, one of the filed with `null` value within the composite keys  is accepted. Actually, none of the field within the composite keys is allowed to be null according to the standard rule of SQL, and this is what we encountered on postgres SQL.

switch => {
        cols => [qw(node switch port vlan interface comments disable)],
        **keys => [qw(node switch port)],**

Maybe we will encounter more problems not only the `switch table` when we switch to the postgres sql. The root cause is the bad programming habit - without error handler ...

**Comment by daniceexi** [comment]
@chenglch Thanks for the digging.

@immarvin is it possible to add certain check in the mkdef cmd to find out the case that missing certain composite key.

**Comment by whowutwut** [comment]
This one should have been resolved in the 2.12.4 release.  @tingtli  Can you assign FVT team to help verify and close out?

**Comment by whowutwut** [comment]
I'm going to close this one, please re-open if issues persist
